### PR TITLE
fix: during live requests, the server returns a cursor for the client to use for cache-busting

### PR DESCRIPTION
**Pull request by KyleAMathews**
This PR is a fix for inconsistencies in caching in http proxying while clients are long-polling. It also adds `public` to our `cache-control` header as that's required by some http proxies in order to cache.

HTTP Proxies don't treat the max-age in cache-control exactly the same way. Some start counting the age of the cache from the *beginning* of the request while others count from the *end* of the request.

This inconsistency makes it difficult to reliably control caching and request collapsing behavior for long-polling requests.

My previous PR in this area https://github.com/electric-sql/electric/pull/1656 made request collapsing work nicely with proxies with the first behavior as they'd collapse all requests within the time from the start of a long-poll and the end of the max-age. And when the client went to request again after the long-poll had ended, the previous request cache had expired already so a new request would get sent to the origin.

However, this approach caused issues with proxies with the second behavior as request collapsing would work but when the client re-polled, the cache hadn't yet expired so the client would go into an infinite loop requesting the same cached response over and over.

So this PR adds a `cursor` generated by the server that clients use as part of `live` requests. This skips by any caches from the previous live request (which on proxies with the first behavior, would have expired already).

The cursor is generated by finding the next alignment boundary. I.e. if the timeout is 20 seconds (which it is now but this could change) then we calculate the alignment boundary by taking the current unix timestamp and subtracting the Electric Epoch of October 9th, 2024 then dividing by 20 and rounding up and the multiplying by 20 again.

In practice this partitions caches for live requests for a given offset into 20 second windows.

### CODE REVIEW

File: website/electric-api.yaml [hunk 1]
```diff
@@ -127,7 +127,7 @@ paths:
             cache-control:
               schema:
                 type: string
-              example: "max-age=60, stale-while-revalidate=300"
+              example: "public, max-age=60, stale-while-revalidate=300"
```
Comment (thruflo) [inline website/electric-api.yaml:130]: Should we also add the cursor param to the openapi spec?

Comment (KyleAMathews) [reply website/electric-api.yaml:130]: ah right — will do

**Review by balegas — APPROVED** [review]
Okay, this will work. Now we can increase the max-age if we want, right? More collapsing, more data at the edge.
It also helps with nextjs caching behaviour, which I was sidestepping by adding a random to the URL... sounds familiar.

I've approved, but might be useful if team has a look at the Elixir before merging.

**Comment by KyleAMathews** [comment]
> Now we can increase the max-age if we want, right

Yup — we could make it configurable. Probably also in practice we'll need to let clients set the long-polling seconds as well as there's some proxies that have pretty short timeouts. But the default could be much higher.

File: packages/sync-service/lib/electric/plug/serve_shape_plug.ex [hunk 2]
```diff
@@ -331,16 +343,20 @@ defmodule Electric.Plug.ServeShapePlug do
 
   defp put_resp_cache_headers(%Conn{assigns: %{config: config, live: live}} = conn, _) do
```
Comment (msfstef) [inline packages/sync-service/lib/electric/plug/serve_shape_plug.ex:344]: this is actually getting called _before_ `:serve_log_or_snapshot`, so the interval it gets assigned to is based off of the time closer to the start of the request - just ensuring this is intentional

**Review by msfstef — COMMENTED** [review]
What an annoying issue from lack of standardisation! I like having a custom cache cursor though for the live requests, relatively cheap to implement and maintain and gives us better control.

I've left some comments and questions for clarification

File: packages/sync-service/lib/electric/plug/serve_shape_plug.ex (hunk 2 above)
Comment (KyleAMathews) [reply packages/sync-service/lib/electric/plug/serve_shape_plug.ex:344]: it doesn't really matter what the time is as long as it's consistent

File: packages/sync-service/lib/electric/plug/serve_shape_plug.ex [hunk 3]
```diff
@@ -23,6 +23,18 @@ defmodule Electric.Plug.ServeShapePlug do
   @up_to_date [Jason.encode!(%{headers: %{control: "up-to-date"}})]
   @must_refetch Jason.encode!([%{headers: %{control: "must-refetch"}}])
 
+  defmodule TimeUtils do
+    def seconds_since_oct9th_2024_next_interval do
+      oct9th2024 = DateTime.from_naive!(~N[2024-10-09 00:00:00], "Etc/UTC")
+      now = DateTime.utc_now()
+
+      diff_in_seconds = DateTime.diff(now, oct9th2024, :second)
+      next_interval = ceil(diff_in_seconds / 20) * 20
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/sync-service/lib/electric/plug/serve_shape_plug.ex:32]: Oh sweet! I didn't know there was a config already for this — I'll switch to it.

File: packages/typescript-client/src/client.ts [hunk 4]
```diff
@@ -153,6 +155,7 @@ export class ShapeStream<T extends Row<unknown> = Row>
     validateOptions(options)
     this.options = { subscribe: true, ...options }
     this.#lastOffset = this.options.offset ?? `-1`
+    this.#nextLiveCursor = ``
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/typescript-client/src/client.ts:158]: yeah the initial request is different so collapsing/caching will be different. I couldn't think of any way around this but it's also not that big of deal as it's just one more request basically getting to Electric.

File: packages/typescript-client/src/client.ts [hunk 5]
```diff
@@ -143,6 +144,7 @@ export class ShapeStream<T extends Row<unknown> = Row>
   >()
 
   #lastOffset: Offset
+  #nextLiveCursor: string // Seconds since our Electric Epoch 😎
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/typescript-client/src/client.ts:147]: true — I'll fix

File: packages/sync-service/lib/electric/plug/serve_shape_plug.ex [hunk 6]
```diff
@@ -331,16 +343,20 @@ defmodule Electric.Plug.ServeShapePlug do
...
+      |> put_resp_header(
+        "electric-next-cursor",
+        TimeUtils.seconds_since_oct9th_2024_next_interval() |> Integer.to_string()
       )
     else
       put_resp_header(
         conn,
         "cache-control",
-        "max-age=#{config[:max_age]}, stale-while-revalidate=#{config[:stale_age]}"
+        "public, max-age=#{config[:max_age]}, stale-while-revalidate=#{config[:stale_age]}"
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/sync-service/lib/electric/plug/serve_shape_plug.ex:359]: Yeah we do want data to be cachable — it's up to people to authorize in front of the caches but it does need to be cachable in general.

File: packages/sync-service/lib/electric/plug/serve_shape_plug.ex [hunk 7]
```diff
@@ -23,6 +23,18 @@ defmodule Electric.Plug.ServeShapePlug do
   @up_to_date [Jason.encode!(%{headers: %{control: "up-to-date"}})]
   @must_refetch Jason.encode!([%{headers: %{control: "must-refetch"}}])
 
+  defmodule TimeUtils do
+    def seconds_since_oct9th_2024_next_interval do
+      oct9th2024 = DateTime.from_naive!(~N[2024-10-09 00:00:00], "Etc/UTC")
```
[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/sync-service/lib/electric/plug/serve_shape_plug.ex:28]: ok good idea as this will be called a ton

**Pull request merged**